### ISSUE TITLE
PPO2 episode reward  drops catastrophically during training

### ISSUE THREAD

**Issue by kp368** [hill-a/stable-baselines#837]
Apologies if this is a more general RL question. I am using PPO2 with default params, with CNN policy. After about 100K I get a perfect model. Later on the model suddenly deteriorates and at the end of training it's useless. I understand that I should always save the best model and not the latest. However, why is this happening? Is it normal/expected? I have observed it with PPO2 in a number of applications.

<img width="1231" alt="Screenshot 2020-05-01 at 13 28 02" src="https://user-images.githubusercontent.com/6249849/80805328-a6ab4000-8baf-11ea-9be1-978f37749d9d.png">

**Comment by Miffyli** [comment]
You could find better answers in [the docs](https://stable-baselines.readthedocs.io/en/master/guide/rl_tips.html) or in [OpenAI SpinningUp](https://spinningup.openai.com/en/latest/).

But TL;DR: Yes, this can happen. It may happen because of mathematical inaccuracies in updates (should be quite ironed out in stable-baselines), or simply due to environment/agent setup. E.g. it learns to complete the task one way (which gets high reward), but because of exploration it attempts something else that also seems promising. Sometimes this might lead outside the good, initial track, leading network to only be trained on the bad samples which may lead to forgetting the initial policy.

**Comment by PartiallyTyped** [comment]
> You could find better answers in [the docs](https://stable-baselines.readthedocs.io/en/master/guide/rl_tips.html) or in [OpenAI SpinningUp](https://spinningup.openai.com/en/latest/).

Note, the spinning up implementation uses the approximate kl divergence as an early stop mechanism. This isn't implemented in stable-baselines ([tf1](https://github.com/openai/spinningup/blob/master/spinup/algos/tf1/ppo/ppo.py#L234-L236), [torch](https://github.com/openai/spinningup/blob/038665d62d569055401d91856abb287263096178/spinup/algos/pytorch/ppo/ppo.py#L269-L271)).

@kp368 
You should know, policy gradients simply approximate the true gradient and with small batch sizes, they are usually far off, thus, they compensate with small lr.  In addition, PPO simply stops gradients from flowing backwards. The gradient of clip for values outside the range is 0. So this is equivalent of using an even smaller batch size. 

Some good reads:
- [Implementation Matters in Deep RL: A Case Study on PPO and TRPO ](https://openreview.net/attachment?id=r1etN1rtPB)
- [Are Deep Policy Gradient AlgorithmsTruly Policy Gradient Algorithms?](https://arxiv.org/pdf/1811.02553.pdf)
- [Deep Reinforcement Learning that Matters](https://arxiv.org/abs/1709.06560)
